graph/src/ipfs_client.rs

```diff
@@ -139,41 +145,42 @@ impl IpfsClient {
         }
     }
 
-    /// Calls stat for the given API route, and returns the total size of the object.
-    pub async fn stat_size(
-        &self,
-        api: StatApi,
-        mut cid: String,
-        timeout: Duration,
-    ) -> Result<u64, reqwest::Error> {
-        let route = format!("{}/stat", api.route());
-        if api == StatApi::Files {
-            // files/stat requires a leading `/ipfs/`.
-            cid = format!("/ipfs/{}", cid);
-        }
-        let url = self.url(&route, &cid);
-        let res = self.call(url, None, Some(timeout)).await?;
-        match api {
-            StatApi::Files => Ok(res.json::<FilesStatResponse>().await?.cumulative_size),
-            StatApi::Block => Ok(res.json::<BlockStatResponse>().await?.size),
-        }
+    /// To check the existence of a cid, we do a cat of a single byte.
+    pub async fn exists(&self, cid: &str, timeout: Option<Duration>) -> Result<(), IpfsError> {
+        self.call(self.cat_url("cat", cid, Some(1)), None, timeout)
+            .await?;
+        Ok(())
     }
 
-    /// Download the entire contents.
-    pub async fn cat_all(&self, cid: &str, timeout: Duration) -> Result<Bytes, reqwest::Error> {
-        self.call(self.url("cat", cid), None, Some(timeout))
-            .await?
-            .bytes()
-            .await
-    }
+    pub async fn cat_all(
+        &self,
+        cid: &str,
+        timeout: Option<Duration>,
+        max_file_size: usize,
+    ) -> Result<Bytes, IpfsError> {
+        let byte_stream = self.cat_stream(cid, timeout).await?;
+        let bytes = byte_stream
+            .err_into()
+            .try_fold(BytesMut::new(), |mut acc, chunk| async move {
+                acc.extend_from_slice(&chunk);
+
+                // Check size limit
+                if acc.len() > max_file_size as usize {
+                    return Err(IpfsError::FileTooLarge(cid.to_string(), max_file_size));
+                }
 
-    pub async fn cat(
+                Ok(acc)
+            })
+            .await?;
+        Ok(bytes.into())
+    }
+    pub async fn cat_stream(
         &self,
         cid: &str,
         timeout: Option<Duration>,
-    ) -> Result<impl Stream<Item = Result<Bytes, reqwest::Error>>, reqwest::Error> {
+    ) -> Result<impl Stream<Item = Result<Bytes, reqwest::Error>> + 'static, reqwest::Error> {
         Ok(self
-            .call(self.url("cat", cid), None, timeout)
+            .call(self.cat_url("cat", cid, None), None, timeout)
```

Comment:
I've looked at the reqwest docs and trawled through the code a bit, but none of it indicates that this will actually read and return the response in chunks. I assume it does, but it's a bit strange that none of the underlying docs guarantee that we don't read everything into memory before getting a response.

Reply:
Reqwest really isn't that well documented. But I'm assuming `bytes_stream` does what we'd expect and gives us partial data as soon as it is available.